File: java-web-spring-boot-3/src/main/java/com/vndevteam/javawebspringboot3/infrastructure/logging/MDCLoggingFilter.java
package com.vndevteam.javawebspringboot3.infrastructure.logging;

import com.vndevteam.javawebspringboot3.infrastructure.config.Slf4jMDCFilterConfig;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.sql.Timestamp;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

@Component
@Slf4j
public class MDCLoggingFilter extends OncePerRequestFilter {
    public static final String MDC_TOKEN_KEY = "RequestFilter.UUID";
    public static final String MDC_CLIENT_IP_KEY = "RequestFilter.ClientIP";

    @Value("${app.logging.log-response-time.enable}")
    Boolean enableMeasureTime = false;

    @Value("${app.logging.log-response-time.exclude}")
    List<String> excludeApiPath = List.of();

    @Value("${app.logging.log-response-time.include}")
    List<String> includeApiPath = List.of();

    private final String responseHeader;
    private final String requestHeader;

    public MDCLoggingFilter() {
        this.responseHeader = Slf4jMDCFilterConfig.DEFAULT_RESPONSE_TOKEN_HEADER;
        this.requestHeader = null;
    }

    public MDCLoggingFilter(String responseHeader, String requestHeader) {
        this.responseHeader = responseHeader;
        this.requestHeader = requestHeader;
    }

    @Override
    protected void doFilterInternal(
            @NonNull HttpServletRequest request,
            @NonNull HttpServletResponse response,
            @NonNull FilterChain filterChain) throws ServletException, IOException {
        Instant start = Instant.now();
        try {
            if (MDC.get(MDC_CLIENT_IP_KEY) == null) {
                MDC.put(MDC_CLIENT_IP_KEY, getClientIpAddress(request));
            }

            if (MDC.get(MDC_TOKEN_KEY) == null && !responseHeader.isBlank()) {
                String clientId = getClientId(request);
                MDC.put(MDC_TOKEN_KEY, clientId);
                response.addHeader(responseHeader, clientId);
            }

            filterChain.doFilter(request, response);
        } finally {
            if (enableMeasureTime
                    && isLoggingRequest(
                    includeApiPath, excludeApiPath, request.getRequestURL().toString())) {
                Instant finish = Instant.now();
                long time = Duration.between(start, finish).toMillis();
                log.info("Request URL: {} - status: {} - Take Time: {} millisecond (start: {}, end: {}).",
                        request.getRequestURL(), response.getStatus(), time, Timestamp.from(start), Timestamp.from(finish));
            }
            MDC.remove(MDC_TOKEN_KEY);
            MDC.remove(MDC_CLIENT_IP_KEY);
        }
    }

    @Override
    protected boolean shouldNotFilterAsyncDispatch() {
        return true;
    }

    private String getClientIpAddress(HttpServletRequest httpHeaders) {
        String xForwardedHeader = httpHeaders.getHeader("X-Forwarded-For");
        if (xForwardedHeader == null || xForwardedHeader.isBlank()) {
            return httpHeaders.getRemoteAddr();
        }
        return xForwardedHeader.split(",")[0].trim();
    }

    private String getClientId(HttpServletRequest request) {
        String clientId;
        if (requestHeader != null
                && !requestHeader.isBlank()
                && request.getHeader(requestHeader) != null
                && !request.getHeader(requestHeader).isBlank()) {
            clientId = request.getHeader(requestHeader);
        } else {
            clientId = UUID.randomUUID().toString();
        }
        return clientId;
    }

    private boolean isLoggingRequest(
            List<String> includeApiPath, List<String> excludeApiPath, String requestUrl) {
        if (includeApiPath.isEmpty() && excludeApiPath.isEmpty()) {
            return true;
        } else if (!includeApiPath.isEmpty() && excludeApiPath.isEmpty()) {
            return includeApiPath.stream().anyMatch(requestUrl::contains);
        } else if (includeApiPath.isEmpty()) {
            return excludeApiPath.stream().noneMatch(requestUrl::contains);
        } else {
            return includeApiPath.stream().anyMatch(requestUrl::contains);
        }
    }
}

File: kotlin-web-spring-boot-3/src/main/kotlin/com/vndevteam/kotlinwebspringboot3/infrastructure/logging/MDCLoggingFilter.kt
package com.vndevteam.kotlinwebspringboot3.infrastructure.logging

import com.vndevteam.kotlinwebspringboot3.infrastructure.config.Slf4jMDCFilterConfig
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import java.sql.Timestamp
import java.time.Duration
import java.time.Instant
import java.util.*
import org.jboss.logging.MDC
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class MDCLoggingFilter : OncePerRequestFilter {
    companion object {
        private const val X_FORWARDED_FOR = "X-Forwarded-For"
        private const val MDC_TOKEN_KEY = "RequestFilter.UUID"
        private const val MDC_CLIENT_IP_KEY = "RequestFilter.ClientIP"
        private val log: Logger = LoggerFactory.getLogger(MDCLoggingFilter::class.java)
    }

    @Value("\${app.logging.log-response-time.enable}") val enableMeasureTime: Boolean = false

    @Value("\${app.logging.log-response-time.exclude}")
    val excludeApiPath: List<String>? = emptyList()

    @Value("\${app.logging.log-response-time.include}")
    val includeApiPath: List<String>? = emptyList()

    private var responseHeader: String? = null
    private var requestHeader: String? = null

    constructor() {
        this.responseHeader = Slf4jMDCFilterConfig.DEFAULT_RESPONSE_TOKEN_HEADER
        this.requestHeader = null
    }

    constructor(
        responseHeader: String,
        requestHeader: String?
    ) {
        this.responseHeader = responseHeader
        this.requestHeader = requestHeader
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val start = Instant.now()
        try {
            if (MDC.get(MDC_CLIENT_IP_KEY) == null) {
                MDC.put(MDC_CLIENT_IP_KEY, getClientIpAddress(request))
            }

            if (MDC.get(MDC_TOKEN_KEY) == null && !responseHeader.isNullOrBlank()) {
                val clientId = getClientId(request)
                MDC.put(MDC_TOKEN_KEY, clientId)
                response.addHeader(responseHeader, clientId)
            }

            filterChain.doFilter(request, response)
        } finally {
            if (
                enableMeasureTime &&
                    isLoggingRequest(
                        includeApiPath!!,
                        excludeApiPath!!,
                        request.requestURL.toString()
                    )
            ) {
                val finish = Instant.now()
                val time: Long = Duration.between(start, finish).toMillis()
                log.info(
                    "Request URL: ${request.requestURL} - Take Time: $time millisecond (start: ${Timestamp.from(start)}, end: ${Timestamp.from(finish)} )."
                )
            }
            MDC.remove(MDC_TOKEN_KEY)
            MDC.remove(MDC_CLIENT_IP_KEY)
        }
    }

    override fun shouldNotFilterAsyncDispatch(): Boolean {
        return true
    }

    private fun getClientIpAddress(httpHeaders: HttpServletRequest): String {
        val xForwardedHeader = httpHeaders.getHeader(X_FORWARDED_FOR)
        if (xForwardedHeader.isNullOrBlank()) {
            return httpHeaders.remoteAddr
        }
        return xForwardedHeader.split(',')[0].trim()
    }

    private fun getClientId(request: HttpServletRequest): String {
        val clientId: String =
            if (
                !requestHeader.isNullOrBlank() && !request.getHeader(requestHeader).isNullOrBlank()
            ) {
                request.getHeader(requestHeader)
            } else {
                UUID.randomUUID().toString()
            }
        return clientId
    }

    private fun isLoggingRequest(
        includeApiPath: List<String>,
        excludeApiPath: List<String>,
        requestUrl: String
    ): Boolean {
        return when {
            includeApiPath.isEmpty() && excludeApiPath.isEmpty() -> {
                true
            }
            includeApiPath.isNotEmpty() && excludeApiPath.isEmpty() -> {
                includeApiPath.any { requestUrl.contains(it) }
            }
            includeApiPath.isEmpty() -> {
                !excludeApiPath.any { requestUrl.contains(it) }
            }
            else -> {
                includeApiPath.any { requestUrl.contains(it) }
            }
        }
    }
}
